upload/src/addons/SV/WarningImprovements/XF/Service/User/Warn.php

```diff
@@ -72,6 +72,23 @@ protected function getCustomWarningDefinition()
             ->fetchOne();
     }
 
+    protected function _save()
+    {
+        $return = parent::_save();
+
+        if ($return instanceof \XF\Entity\Warning)
```

Comment:
Should rename `$return` to `$warning` , and compare to the extended warning type (`\SV\WarningImprovements\XF\Entity\Warning`).

I'm fairly sure, you should use `getAnonymizedIssuer()` not `WarnedBy` here. As otherwise the sender username will leak in two ways;
- the sender username/avatar is written into the alert envelop that the per-alert template doesn't touch.
- on upgrading the add-on, the add-on is put into a half-disabled state where templates work but any code extension is broken.

Reply:
@Xon I've made some more changes could you give it a try? Made some changes

Reply:
LGTM